packages/format-library/src/link/inline.js

```diff
@@ -148,14 +151,37 @@ function InlineLinkUI( {
 					newText.length
 				);
 
+				// Get the boundaries of the active link format.
+				const boundary = getFormatBoundary( value, {
+					type: 'core/link',
+				} );
+
+				// Split the value at the start of the active link format.
+				// Passing "start" as the 3rd parameter is required to ensure
+				// the second half of the split value is split at the format's
+				// start boundary and avoids relying on the value's "end" property
+				// which may not correspond correctly.
+				const [ valBefore, valAfter ] = split(
+					value,
+					boundary.start,
+					boundary.start
```

Comment:
@ellatrix I don't think passing this should be necessary but it is.

Firstly the `split()` function only accepts 2 args not 3. The only reason this works is that it uses `...arguments` to delegate to the `splitAtSelection` function which _does_ accept a 3rd parameter.

By default the 3rd param of `splitAtSelection` defaults to the passed `value`'s `end` property. In the case of `split()` we always want to split the string at position `X` as defined by the 2nd argument. However in some cases the `start` and `end` of the `value` may not match and thus the `value` is split incorrectly.

Reply:
Yes, by default all of these functions look at the rich text value's current selection. We allow you to override that, and to do that properly you need to provide both a new selection start and end.

Reply:
...but `split()` doesn't actually have an `end` argument, it just has `string`

https://github.com/WordPress/gutenberg/blob/9e6e8c9cb7e8cf35ea148d28dfb7dc21b1c4b915/packages/rich-text/src/split.js#L19

https://github.com/WordPress/gutenberg/blob/9e6e8c9cb7e8cf35ea148d28dfb7dc21b1c4b915/packages/rich-text/src/split.js#L15

The only reason it works is that `...arguments` is utilised:

https://github.com/WordPress/gutenberg/blob/9e6e8c9cb7e8cf35ea148d28dfb7dc21b1c4b915/packages/rich-text/src/split.js#L20-L22

So it's not obvious from the function signature that you can/should provide an `end`. 

> We allow you to override that, and to do that properly you need to provide both a new selection start and end.

Is the intention that you manually update the selection on the `value` being passed into `split` before it's passed in?


I still feel it's confusing...

`endIndex` is set to `value.end` which may produce unexpected results when (as instructed in the doc block) you've passed an explicit `Start index` into the `split()` as the 2nd `string` argument. 

https://github.com/WordPress/gutenberg/blob/9e6e8c9cb7e8cf35ea148d28dfb7dc21b1c4b915/packages/rich-text/src/split.js#L57-L61

If I pass X as `string` then I'd expect the value to be split at X. What can actually happen is the value is split at X but the (remaining) second half of the value is _not_ split at X but at whatever `value.end` is.

https://github.com/WordPress/gutenberg/blob/9e6e8c9cb7e8cf35ea148d28dfb7dc21b1c4b915/packages/rich-text/src/split.js#L71-L77

Could / should we instead default `endIndex` to `startIndex` so that (by default) if a start index is passed to `split` as `string` then it will split where you expect it to?